### PR TITLE
Port environments from Pester

### PR DESCRIPTION
**Pull request by apurvis**
@slpsys i powered through this pretty quickly; you see any obvious errors?

p.s. minitest totally sucks.

### CODE REVIEW

**Comment by apurvis** [comment]
(i didn't open a PR on the main repo yet; wanted to be sure we feel good about the feature before we request a merge to master)

File: lib/retriable.rb [hunk 1]
```diff
@@ -70,4 +72,16 @@ def retriable(opts = {})
       end
     end
   end
+
+  def respond_to?(method_sym, options = {}, &block)
+    super || config.environments.key?(method_sym)
+  end
```
Comment (slpsys) [inline lib/retriable.rb]: dunno if they're going to yell about this; i think i copied it from a blog post. i understand why and what it does, but dunno if it's currently the best practice

File: lib/retriable/environment.rb [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    def initialize(opts)
+      raise ArgumentError, 'Environment.new requires a hash' unless opts.is_a?(Hash)
+      opts.each do |k, v|
+        raise ArgumentError, "#{k} => #{v} is not a valid configuration" unless Config::PROPERTIES.include?(k)
+      end
+
+      @options = opts
+    end
+
+    def method_missing(name, *args, &block)
```
Comment (slpsys) [inline lib/retriable/environment.rb]: AFAIK rubyists prefer that `method_missing` is always paired with `respond_to?`, which is why they're both in `lib/retriable.rb`

File: retriable.gemspec [hunk 3]
```diff
@@ -6,7 +6,7 @@ require "retriable/version"
 Gem::Specification.new do |spec|
   spec.name          = "retriable"
   spec.version       = Retriable::VERSION
-  spec.authors       = ["Jack Chu"]
+  spec.authors       = ["Jack Chu", 'Lumos Labs, Inc.']
```
Comment (slpsys) [inline retriable.gemspec]: unless this is just for Our Official Fork, IDK how they're going to feel about this

**Comment by slpsys** [comment]
As I said in one of the comments, I don't know if this we plan on merging this upstream--I hope so, it's really the motivating feature of [Pester](https://github.com/lumoslabs/pester)--but if so, we may want to roll back that Gemspec change. Otherwise, 🍻

**Comment by apurvis** [comment]
yeah my goal is open a PR onto master once we think this looks reasonable.

File: lib/retriable/environment.rb (hunk 2 above)
Comment (apurvis) [inline lib/retriable/environment.rb]: yeah i didn't know whether it was necessary here or not...  i also am not sure the best way to go about this.

**Comment by apurvis** [comment]
ok i cleaned this up - we don't need to use `method_missing` in `Environment`; only in `Retriable`.  Going to squash merge this to our fork and open a PR to master and see what the feedback is.

**Comment by apurvis** [comment]
https://github.com/kamui/retriable/pull/25

**Pull request merged**